flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -360,8 +360,9 @@ public boolean notifyBufferAvailable(Buffer buffer) {
 
 			// Important: double check the isReleased state inside synchronized block, so there is no
 			// race condition when notifyBufferAvailable and releaseAllResources running in parallel.
-			if (isReleased.get() || bufferQueue.getAvailableBufferSize() >= numRequiredBuffers) {
+			if (isReleased.get() || isBlocked() || bufferQueue.getAvailableBufferSize() >= numRequiredBuffers) {
```

Comment:
That is the exact issue I pointed above. 

I think it will not cause credit leak currently.  But for further improvement, we can compare the current available credits with backlog after unblocking the channel. If not enough, we can trigger to request floating buffers at once. In current implementation, it has to wait next `onSenderBacklog` to trigger this process.

Reply:
Indeed, the "blocked" credits should be in `numRequiredBuffers` which was updated with the last buffer that we received. Any update on that will come with the next buffer. We just need to make sure that we acknowledge the newly freed credit to the sender (the backlog may be 0).

The request for floating buffers should remain fair among all channels though and therefore we cannot request all floating buffers to satisfy that need at once!